Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidgetRepresentation2D.cxx

```diff
@@ -288,13 +400,163 @@ void vtkSlicerMarkupsWidgetRepresentation2D::UpdateAllPointsAndLabelsFromMRML(do
+        double labelTextWidth = labelCharWidth * labelLength;
+        double labelTextHeight = (this->MarkupsDisplayNode->GetTextScale() * 3.) * numberLines;
+
+        // center of background volume in display XY coordinates
+        double backgroundVolumeCenterXY[2] = { 0.0 };
+        this->GetBackgroundVolumeXYCenter(backgroundVolumeCenterXY);
+
+        double leadersLineOffset[2] = { 0.0 };
+
+        this->GetLabelTextXYOffsets(backgroundVolumeCenterXY, slicePos, leadersLineOffset);
+
+        // Add a function to determine the exact bounds of the label text in 2D view
+        double lineXSpacer = labelCharWidth;
+
+        // set connector lineStartPos x and y according to location of fiducial
+        lineStartPos[0] = slicePos[0];
+        lineStartPos[1] = slicePos[1];
+
+        // set lineEndPos and new labelPos according to location of fiducial
+
+        if (slicePos[0] < backgroundVolumeCenterXY[0])
+          {
+          lineEndPos[0] = slicePos[0] + labelsOffset - leadersLineOffset[0];
+          labelPos[0] = lineEndPos[0] - labelTextWidth - lineXSpacer;
+          }
+        else
+          {
+          lineEndPos[0] = slicePos[0] + labelsOffset + leadersLineOffset[0];
+          labelPos[0] = lineEndPos[0] + (lineXSpacer / 2.);
+          }
+
+        if (slicePos[1] < backgroundVolumeCenterXY[1])
+          {
+          lineEndPos[1] = (slicePos[1] + labelsOffset) - leadersLineOffset[1];
+          labelPos[1] = lineEndPos[1] - labelTextHeight;
+          }
+        else
+          {
+          lineEndPos[1] = (slicePos[1] + labelsOffset) + leadersLineOffset[1];
+          labelPos[1] = lineEndPos[1] - (labelTextHeight/ 2.);
+          }
+
+        // handle label text touching slice view borders
+
+        // left screen border
+        if (labelPos[0] < 0.)
+          {
+          labelPos[0] = 0.;
+          lineEndPos[0] = labelPos[0] + labelTextWidth + lineXSpacer;
+          }
+        // right screen border
+        if (labelPos[0] > screenWidth - labelTextWidth)
+          {
+          labelPos[0] = screenWidth - labelTextWidth;
+          lineEndPos[0] = labelPos[0] - lineXSpacer;
+          }
+        // upper screen border
+        if (labelPos[1] > screenHeight - labelTextHeight * 2.)
+          {
+          labelPos[1] = screenHeight - labelTextHeight * 2.;
+          lineEndPos[1] = labelPos[1] + labelTextHeight;
+          }
+        // lower screen border
+        if (labelPos[1] < 0.)
+          {
+          labelPos[1] = 0.;
+          lineEndPos[1] = labelPos[1] + labelTextHeight;
+          }
+
+        pts->InsertPoint(ptcnt, lineStartPos[0], lineStartPos[1], 0.);
+        pts->InsertPoint(ptcnt + 1, lineEndPos[0], lineEndPos[1], 0.);
+        line->InsertNextCell(2);
+        line->InsertCellPoint(ptcnt);
+        line->InsertCellPoint(ptcnt + 1);
+        ptcnt += 2;
+
+        // make the connector line color a bit lighter then the selected glyph/text color
+        double color[3] = { 0,0,0 };
```

Comment:
Do not do all these operations for each point, but do these only once, after all the points are added.